### ISSUE TITLE
Performance issues on Point in Polygon queries

### ISSUE THREAD

**Issue by racevedoo** [tidwall/tile38#291]
Hi,
I'm running some benchmarks regarding tile38 point-in-polygon query performance and getting some strange results. I wonder if I'm doing something awkward or these response times are expected:

Dataset: 400k multipolygons representing countries, states, cities all over the world - total ~9GB
Query: `intersects world ids object {"type":"Point","coordinates":[<random_lng>,<random_lat>}`
Load: 100 concurrent clients, each running 1000 queries
Server specs: AWS c4.2xlarge(8 vCPU, 15GB RAM)
Results:
```
percentile:  50  value:  480.110667 ms
percentile:  80  value:  745.076178 ms
percentile:  90  value:  1083.193823 ms
percentile:  95  value:  1423.035608 ms
percentile:  99  value:  2046.497578 ms
percentile:  99.9  value:  2724.5885485 ms
percentile:  99.99  value:  3234.4719619999996 ms
```

Can anyone help me?
Thanks in advance!

**Comment by tidwall** [comment]
This may be related to #267 where multipolygons, and other geojson object that are made up of more than one geometries, are stored as a one rectangle in the r-tree index instead of storing each child geometry with it's own rectangle.

It's hard to say if this is absolutely the same issue that you're running into but it does sound similar.

Are you using v1.11.1? I recently added some optimizations that speeds up queries a little bit on large complex multipolygons.

**Comment by racevedoo** [comment]
Yes, I'm using v1.11.1
I checked #267 before opening this.
I think that what makes the performance deteriorate is the fact that the polygons are way too big(countries, for example) . What do you think? Any ideas to mitigate this?

**Comment by tidwall** [comment]
Sorry for the delayed response. 

Yes, very big polygons can slow performance quite a bit. This seems to be more common with shapes like countries. I've been testing various optimizations over the past few weeks and I hope to speed things up soon.

**Comment by tidwall** [comment]
Yes, please leave open for now. Thanks :)

**Comment by abhit011** [comment]
@racevedoo

Just curious to know , how many tile38 nodes you are running ?

**Comment by amirmasud** [comment]
Any update guys?

**Comment by racevedoo** [comment]
@abhit011 
Sorry for keeping you waiting, completely forgot to answer! I was running a single node, as i mentioned in the issue:
> c4.2xlarge(8 vCPU, 15GB RAM)

**Comment by racevedoo** [comment]
@amirmasud 
I don't know, maybe @tidwall ?

**Comment by tidwall** [comment]
I haven't yet released any major updates point-in-polygon enhancements. I've experimented with a couple different options, all of which include breaking up geometries into smaller components, but it's still a work in progress. It's absolutely still on the radar and a high priority. 

I'll keep this issue updated as it progresses.

**Comment by amirmasud** [comment]
@tidwall  As an experiment, We subdivided large polygons like provinces and cities into 8 vertex polygons (in postgis) and then we inserted them into the tile38. After applying this trick, our point in polygon's load test results showed an impressive improvement. (about 10x load bearing)

Maybe it's a notable idea to subdivide extra-large polygons (or ones with hundreds of vertices) to multiple smaller polygons on insert time to achieve better performance on point in polygon.

**Comment by tidwall** [comment]
@amirmasud Thanks for the info. I've been seeing very good results in my tests too. I hoping to add something like that to Tile38 very soon.

**Comment by tidwall** [comment]
A quick update.

I've been able to successfully modify the geo library to support deep indexing on individually complex polygons and line strings. I haven't yet plugged everything into Tile38, so the final results may vary, but what I'm seeing so far looks promising.

Here I'm performing 1M random point-in-polygon operations on the exterior boundary of the lower 48 states (61K point polygon)

```
No index: 1,000,000 ops over 6 threads in 41217ms, 24,261/sec, 41217 ns/op
Index:    1,000,000 ops over 6 threads in 348ms, 2,873,853/sec, 347 ns/op
```

Rather than subdividing the polygon into smaller parts, I'm simply loading the segments into a discrete rtree and raycasting the point as a rect with infinite X extents.

These are results on my machine when I'm hitting the new library directly. 

I'll notify this thread when it's fully integrated into Tile38.

**Comment by racevedoo** [comment]
@tidwall the results look great! Looking forward to having it all integrated into Tile38!

**Comment by tidwall** [comment]
I just pushed the performance upgrade to the master branch. 

This enhancement includes a lot of changes to the overall codebase, including splitting out the GeoJSON package into it's own [repo](https://github.com/tidwall/geojson). Pretty much every geometry calculation has been overhauled and complex geometries now include micro indexing.

It should be fully compatible with the previous version.

One side effect could be that some polygons may use more memory. So please look out for that and anything else you stumble upon.
 
Thanks!

**Comment by tidwall** [comment]
I'm closing this issue because it appears to be resolved with the changes in the v1.14.0 release.
